### ISSUE TITLE
BLOB Texts to string on NodeJS

### ISSUE THREAD

**Issue by turalgulmammadov** [hgourvest/node-firebird#82]
Hey guys,

I have so much troubles to convert BLOB Texts to string on nodejs. Can you please show me an easy way? Thanks
P.S I can convert the BLOB images to actual images but the method that I am using on there is not working for texts.

**Comment by turalgulmammadov** [comment]
@mariuz 
@sdnetwork 
Any suggestions guys?

**Comment by tomcon** [comment]
Is the character set of your db UTF?

**Comment by turalgulmammadov** [comment]
@tomcon yes. I have tried the code below as well. Even it's not working:
```
pool.get(function(err, db) {
    if (err)
        throw err;
    var id = 0;
    var k = 0;
    db.sequentially('SELECT a.STID, a.STAB FROM TESTING a WHERE a.STAB IS NOT NULL', function(row, index) {
        console.log(row);
        row.stab(function(err, name, e) {
            
            if (err){
                console.log(err);
                throw err;
            }
            e.on('data', function(chunk) {

            });

            e.on('end', function() {

                
            });

        });
    });
    k = k + 1;
}, function(err) {
    if (err)
        throw err;
});
pool.destroy();
```

And my stored procedure **TESTING** code :
```
SET TERM ^ ;
ALTER PROCEDURE TESTING
RETURNS (
    STID Integer,
    STAB Blob sub_type 1 )
AS
BEGIN
  FOR SELECT
            
            STUD_ABSENCE."StudentId",
           LIST(STUD_ABSENCE."AbsenceId")
  
    FROM	STUD_ABSENCE
    GROUP BY STUD_ABSENCE."StudentId", STUD_ABSENCE."EnrolmentId"
   INTO          
                :STID ,
                :STAB

            do
  BEGIN
    SUSPEND;
  END
END^
SET TERM ; ^

```

**NOTE** : It's working well on FlameRobin.

**Comment by tomcon** [comment]
From the notes you seem  to be missing this:

  e.on('end', function() {
                // end reading
                // IMPORTANT: close the connection
                db.detach();
            });

**Comment by tomcon** [comment]
What version of Firebird are you using? 

At what point do you see the  [Error: invalid BLOB ID]. error:
e.on( 'data' ) or earlier?

**Comment by turalgulmammadov** [comment]
@tomcon Actually, I think the example in note is not for pooling. However, I have tried the same example as well. It gives me the same error. 
And I used to the previous last one when I got the error, now I have updated it and still the same error. Yes it comes earlier. 
```
row.stab(function(err, name, e) {
            // console.log('test1');
            if (err){
                console.log(err);
                throw err;
            }
// console.log('test2');
            e.on('data', function(chunk) {
            });
// console.log('test3');
            e.on('end', function() {
            });
        });
```
As you see the code above, if I will uncomment console logs only the first one will work.

**Comment by turalgulmammadov** [comment]
@tomcon Actually, regular blob works fine. I need the results of **LIST()** call. Because, I need to concatenate the things.

**Comment by tomcon** [comment]
LIST() seems to return a string
http://www.firebirdsql.org/refdocs/langrefupd21-aggrfunc-list.html

Try CASTing the LIST() to a blob 
and/or concatenating the result of the LIST() with a string '  ', then cast
http://www.firebirdsql.org/refdocs/langrefupd21-intfunc-cast.html
"Casting BLOBs: Successful casting to and from BLOBs is possible since Firebird 2.1."

**Comment by turalgulmammadov** [comment]
It seems that it returns TEXTs as a **BLOB** not string
```
LIST()

Available in: DSQL, PSQL

Added in: 2.1

Description: LIST returns a string consisting of the non-NULL argument values in the group, separated either by a comma or by a user-supplied delimiter. If there are no non-NULL values (this includes the case where the group is empty), NULL is returned.

Result type: BLOB

Syntax: 

LIST ([ALL | DISTINCT] expression [, separator])
ALL (the default) results in all non-NULL values to be listed. With DISTINCT, duplicates are removed, except if expression is a BLOB.

The optional separator argument may be a string literal, a parameter or a variable in versions up to 2.1.3. Starting at 2.1.4 it may be any string expression. This makes it possible to specify e.g. ascii_char(13) as a separator.

The expression and separator arguments support BLOBs of any size and character set.

Date/time and numerical arguments are implicitly converted to strings before concatenation.

**The result is a text BLOB, except when expression is a BLOB of another subtype.**

The ordering of the list values is undefined.
```

**Comment by turalgulmammadov** [comment]
@flakekun Yes I can do but my **BLOB**s will be longer than the MAX that VARCHAR can store. Any other suggestions?

**Comment by malpacasville** [comment]
Seems like this is the same problem described in issue #51

**Comment by sdnetwork** [comment]
try that it should fix the problem 

https://github.com/sdnetwork/node-firebird

**Comment by johnjaider1000** [comment]
I had the same problem and the solution I gave was the following:

SELECT CAST(CAST(a.NEW_INFO  AS BLOB SUB_TYPE TEXT CHARACTER SET utf8) AS VARCHAR(8191) CHARACTER SET utf8) AS NEW_INFO FROM AUDITORIA a;

Change the name of the table (AUDITORIA) and (NEW_INFO) in the query for your field and that's it.

**Comment by mariuz** [comment]
Resolved in the 2.x line: "invalid BLOB ID" came from reading a blob after its transaction was gone — blob readers are now bound to the transaction that produced them, and for text blobs you can skip the reader entirely with `blobAsText: true` (plain strings in the row). Docs: [Reading Blobs](https://github.com/hgourvest/node-firebird#reading-blobs-asynchronous) and the [FAQ](https://github.com/hgourvest/node-firebird#faq). Closing.
